File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        pref = [1]
        suff = [1]

        prod = nums[0]

        for i in range(1,len(nums)):
            pref.append(prod)
            prod *= nums[i]

        prod = nums[-1]
        for i in range(len(nums)-2,-1,-1):
            suff.append(prod)
            prod *= nums[i]


        for i in range(len(nums)):
            nums[i]=pref[i]*suff[len(nums)-i-1]
        print(pref)
        print(suff)
        return nums
        

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        // TC --> O(N)
        // SC --> O(1)
        
        int n = nums.size();
        vector<int> answer(n);

        int product = 1;
        for(int i=0; i < n; i++) {
            answer[i] = product;
            product*=nums[i];
        }

        product = 1;
        for(int i=n-1; i >= 0; i--) {
            answer[i]*=product;
            product*=nums[i];
        }

        return answer;
    }
};